File: problems/anagram-detection/anagram-detection.php
<?php

function hashString($str)
{
  $hashMap = [
    'a' => 2,
    'b' => 3,
    'c' => 5,
    'd' => 7,
    'e' => 11,
    'f' => 13,
    'g' => 17,
    'h' => 19,
    'i' => 23,
    'j' => 29,
    'k' => 31,
    'l' => 37,
    'm' => 41,
    'n' => 43,
    'o' => 47,
    'p' => 53,
    'q' => 59,
    'r' => 61,
    's' => 67,
    't' => 71,
    'u' => 73,
    'v' => 79,
    'w' => 83,
    'x' => 89,
    'y' => 97,
    'z' => 101,
    'A' => 103,
    'B' => 107,
    'C' => 109,
    'D' => 113,
    'E' => 127,
    'F' => 131,
    'G' => 137,
    'H' => 139,
    'I' => 149,
    'J' => 151,
    'K' => 163,
    'L' => 167,
    'M' => 173,
    'N' => 179,
    'O' => 181,
    'P' => 191,
    'Q' => 193,
    'R' => 197,
    'S' => 199,
    'T' => 211,
    'U' => 223,
    'V' => 227,
    'W' => 229,
    'X' => 233,
    'Y' => 239,
    'Z' => 241
  ];

  $hash = 1;
  foreach (str_split($str) as $char)
  {
    $hash *= $hashMap[$char];
  }
  return $hash;
}

function anagramDetection($parent, $child)
{
  $len = strlen($child);
  $childHash = hashString($child);
  $matches = [];

  for ($start = 0; $start < strlen($parent) - $len; $start++)
  {
    $parentSubstr = substr($parent, $start, $len);
    $parentHash = hashString($parentSubstr);
    if ($parentHash === $childHash)
    {
      $matches[] = $parentSubstr;
    }
  }
  return $matches;
}

print_r(anagramDetection('AdnBndAndBdaBn', 'dAn')); // 4 ("Adn", "ndA", "dAn", "And")

File: problems/anagram-detection/anagram-detection.kt
private val charToPrime: Map<Char, Int> = mapOf(
    'a' to 2,
    'b' to 3,
    'c' to 5,
    'd' to 7,
    'e' to 11,
    'f' to 13,
    'g' to 17,
    'h' to 19,
    'i' to 23,
    'j' to 29,
    'k' to 31,
    'l' to 37,
    'm' to 41,
    'n' to 43,
    'o' to 47,
    'p' to 53,
    'q' to 59,
    'r' to 61,
    's' to 67,
    't' to 71,
    'u' to 73,
    'v' to 79,
    'w' to 83,
    'x' to 89,
    'y' to 97,
    'z' to 101,
    'A' to 103,
    'B' to 107,
    'C' to 109,
    'D' to 113,
    'E' to 127,
    'F' to 131,
    'G' to 137,
    'H' to 139,
    'I' to 149,
    'J' to 151,
    'K' to 163,
    'L' to 167,
    'M' to 173,
    'N' to 179,
    'O' to 181,
    'P' to 191,
    'Q' to 193,
    'R' to 197,
    'S' to 199,
    'T' to 211,
    'U' to 223,
    'V' to 227,
    'W' to 229,
    'X' to 233,
    'Y' to 239,
    'Z' to 241
)

fun String.hash(): Int = map { charToPrime[it] ?: error("Unmapped char $it") }.fold(1) { acc, v -> acc * v }

fun anagramDetection(parent: String, child: String): List<String> = buildList {
    when {
        parent.isEmpty() || child.isEmpty() || parent.length < child.length -> {}
        parent.length == child.length -> child.takeIf { parent.hash() == it.hash() }?.run(::add)
        else -> {
            val childHash: Int = child.hash()
            parent.windowed(child.length, 1)
                .filter { it.hash() == childHash }
                .run(::addAll)
        }
    }
}

check(anagramDetection("", "") == emptyList<String>())
check(anagramDetection("x", "foo") == emptyList<String>())
check(anagramDetection("foo", "bar") == emptyList<String>())
check(anagramDetection("AdnBndAndBdaBn", "dAn") == listOf("Adn", "ndA", "dAn", "And"))
